src/main/drivers/dshot.c

```diff
@@ -184,92 +187,90 @@ static uint32_t dshot_decode_eRPM_telemetry_value(uint16_t value)
     return (1000000 * 60 / 100 + value / 2) / value;
 }
 
-static void dshot_decode_telemetry_value(uint8_t motorIndex, uint32_t *pDecoded, dshotTelemetryType_t *pType)
+static void dshotDecodeAndSetTelemetryValue(uint32_t *pDecoded, uint16_t *pHighByte, uint8_t motorIndex, uint16_t value, debugType_e debugType)
 {
-    uint16_t value = dshotTelemetryState.motorState[motorIndex].rawValue;
-    const unsigned motorCount = motorDeviceCount();
+    *pDecoded = value & DSHOT_TELEMETRY_VALUE_MASK;
+
+    // Update debug buffer (motorIndex < motorCount guaranteed by caller)
+    if (motorIndex < DEBUG16_VALUE_COUNT) {
+        *pHighByte = dshotTelemetryState.motorState[motorIndex].telemetryData[DSHOT_TELEMETRY_TYPE_STATUS] << SHIFT_BYTE;
+        DEBUG_SET(debugType, motorIndex, *pHighByte | *pDecoded);
+    }
+}
+
+static void dshotDecodeTelemetryValue(uint8_t motorIndex, uint32_t *pDecoded, dshotTelemetryType_t *pType)
+{
+    uint16_t dshotDebugHighByte;
+    const uint16_t value = dshotTelemetryState.motorState[motorIndex].rawValue;
 
     if (dshotTelemetryState.motorState[motorIndex].telemetryTypes == DSHOT_NORMAL_TELEMETRY_MASK) {   /* Check DSHOT_TELEMETRY_TYPE_eRPM mask */
```

Comment:
This seems to be a bug - telemetryTypes starts as zero, so it is handled as as extended request. It is matches one of extended telemetry values (bit8 is zero, bit 9-11 match one of DSHOT_TELEMETRY_RANGE_), it will be decoded as extended telemetry type an corresponding bit set. That situation is sticky - some erpm values would be decoded as DSHOT_TELEMETRY_RANGE_ until FC reset.

```suggestion
    if ((dshotTelemetryState.motorState[motorIndex].telemetryTypes & DSHOT_TELEMETRY_TYPE_STATE_EVENTS) == 0) { 
```
But that is ugly and fragile. There should be extra flag that is set after enabling extended telemetry

Reply:
regarding the suggestion, since this line does not contain any words like `flag` or `mask`, maybe retain the comment reassuring `bitwise &` usage.